File: lib_common/src/main/java/com/fish/lib_common/base/frame/BasePresenter.java
package com.fish.lib_common.base.frame;

import androidx.annotation.Nullable;

import java.lang.ref.WeakReference;

public abstract class BasePresenter<V extends BaseActivity, M extends BaseModel> {

    protected WeakReference<V> mView;
    protected WeakReference<M> mModel;

    public BasePresenter(V view) {
        mView = new WeakReference<>(view);
        mModel = new WeakReference<>(getModel());
    }

    protected abstract M getModel();

    protected void showLoading(@Nullable String msg) {
        if (mView != null && mView.get() != null) {
            mView.get().showLoading(msg);
        }
    }

    protected void detach() {
        if (mView != null && mView.get() != null) {
            mView.clear();
            mView = null;
        }

        if (mModel != null && mModel.get() != null) {
            mModel.get().detach();
        }
    }
}

File: lib_common/src/main/java/com/fish/lib_common/base/frame/BasePresenter.kt
package com.fish.lib_common.base.frame

import java.lang.ref.WeakReference

@Suppress("FINITE_BOUNDS_VIOLATION_IN_JAVA")
abstract class BasePresenter<V : BaseSimpleActivity, M : BaseModel>(view: V) {

    protected var mView: WeakReference<V>? = null
    protected var mModel: M? = null

    init {
        mView = WeakReference(view)
        mModel = getModel()
    }

    var showLoading = { msg: String?, isShow: Boolean ->
        if (mView != null && mView!!.get() != null) {
            mView?.get()?.showLoading(msg, isShow)
        }
    }

    abstract fun getModel(): M?

    fun detach() {
        if (mView != null && mView!!.get() != null) {
            mView!!.clear()
            mView = null
        }
        mModel?.detach()
        mModel = null
    }
}
